### DISCUSSION TITLE
ERROR: The signature is invalid.

### DISCUSSION THREAD

**Discussion by Richard2012** [spatie/laravel-webhook-client#219] [Q&A] [answered]
We are not certain what is really meant as '**signature_header_name**'  in the config file. This is the likely cause of the 'The signature is invalid' error. How  and where do I find the signature_header_name?

**Comment by PaulUniHomes** [comment, marked as answer]
That would depend on the name of the header that is being sent across in the request to your endpoint. 

If you look in the example, you'll see it is set as 'Signature':

```
$webhookConfig = new \Spatie\WebhookClient\WebhookConfig([
    'name' => 'webhook-sending-app-1',
    'signing_secret' => 'secret-for-webhook-sending-app-1',
    'signature_header_name' => 'Signature', // This is where it is set
    'signature_validator' => \Spatie\WebhookClient\SignatureValidator\DefaultSignatureValidator::class,
    'webhook_profile' => \Spatie\WebhookClient\WebhookProfile\ProcessEverythingWebhookProfile::class,
    'webhook_response' => \Spatie\WebhookClient\WebhookResponse\DefaultRespondsTo::class,
    'webhook_model' => \Spatie\WebhookClient\Models\WebhookCall::class,
    'process_webhook_job' => '',
]);
```

If you look at DefaultSignatureValidator, it pulls the signature from the request headers based on the value you set there:

`$signature = $request->header($config->signatureHeaderName);`

So in this instance, it will look for $signature = $request->header('Signature'); (If you have a look at the WebhookConfig class, you can see how the properties being passed into it as an array are converted).

If you check the headers on the inbound request, you should see if that is what the key for the signature being sent is, and if not, then change the signature_header_name value to whatever the key is in the header.

**Reply by Richard2012** [reply]
Thanks. All clear. What they want here is the **key** in the header which points to the secret **value** (which comes from .env).
The documentation could have been more verbose.

**Reply by Richard2012** [reply]
Sorry guys, I said all clear, but it is not.
After 2 more days of work I found that $config->signatureHeaderName contains a key (from "key:value" pair) to the signature.
In the header itself the value stored under this key comes as some kind of hash, unique to each request.
The SECRET from the .env never comes in the header in my case (Paypal).

Now, I assume the spatie is supposed to verify wheather the hash from the header corresponds with the SECRET value stored in the .env.

But this is where it fails. It returns false.

Here are the real values which should return TRUE during verification (but spatie returns false)
**.env (invalidated after this post):**
`WEBHOOK_CLIENT_SECRET="ENV10VjX3YfNUIWZLeiRMDhXU_PucAVzmTYpC9UHbrr92M9JY_mfKRNmherFPyCwnZ4KKuUTscHzuDMh"`

**webhook-client.php:**
` 'signature_header_name' => 'paypal-transmission-sig',`

**Values from header under paypal-transmission-sig key (2 examples):**

```
BUrgLte0te2571OEq3oapo7ro42+o9RyRGgI82S8rNIhYDumcIFWwhS5BUmFH0Xon6x7p5rxgn1X678R0DURUICO+crLcn3xjB3L0wa2milKOzWqL+xH4rdsMbLQy6vdqbX9Fb2S1uz8zkwLxJQLAZr+G0hBq7AH4AxoGhusabC2RG3dUGs4MLpmIAhTNwrlmFFRGuxzIGS3GaTKYzdzNYHHgO2HhgTfX20VjKL0p1Ldp8yEHHxyYYV8MnpmUHZW63WTUAN62d0ociwiF\/kkrvq\/3J9VEV+t+j0l4osN0kA0vU66Spezpzl\/0W+Hy9vn9QVUujhkFJKauCFIqDU4YQ==

j5mPXQH6BJ8Ir832cmiSZyxvjw1PXc5TjFfGGQRsZtbQ9IyH65iQjHGow25dJX44AhWnhBAHOGamaB1Qz8FdltSNdpkdFoFou89VnflvEo4U5OuueRHOow2M285eiHmur8JAsRBQwwpAQ\/m+VG0I7QPlQHoVUP7MnIo4bHjOKasMwHKrpMWvLuFo8KewdgoKYf2+TDP0JyxHwQ61Q+TKMzXxsQO8fSH36GXgXK8lWg8fViiSPY6CrzMLExHzZAIcCZujXIVZsoYPgEJgW+qyWWJvZYyVOH25JEmHNQg1fFfTrtfcnZE5bUdApH10Q\/911NIWRAzzO1vVADKyHE\/7TA==

```
**Is there annyone at spatie who could reply?**
Thank you
Richard

**Reply by Richard2012** [reply]
The problem is in 
`        $this->ensureValidSignature();
`in WebhookProcessor.php

**Reply by PaulUniHomes** [reply]
Hi Richard

_The SECRET from the .env never comes in the header in my case (Paypal)._

No that's not how the signature works, it's not sending you the signature in a raw format as that would be insecure. I haven't worked with the PayPal webhook, but let me try and explain what I think it might do to help you progress things.

Here's a really basic example:

$signature = $request->header($config->signatureHeaderName);

$signingSecret = $config->signingSecret;

$computedSignature = hash_hmac('sha256', $request->getContent(), $signingSecret);

return hash_equals($signature, $computedSignature);


So on your configuration for the webhook, you'll have something like:
'signing_secret' => env('WEBHOOK_CLIENT_SECRET'),
'signature_header_name' => 'paypal-transmission-sig',

So if we raw out the original basic example:
$signature = 'BUrgLte0te2571OEq3oapo7ro42+o9RyRGgI82S8rNIhYDumcIFWwhS5BUmFH0Xon6x7p5rxgn1X678R0DURUICO+crLcn3xjB3L0wa2milKOzWqL+xH4rdsMbLQy6vdqbX9Fb2S1uz8zkwLxJQLAZr+G0hBq7AH4AxoGhusabC2RG3dUGs4MLpmIAhTNwrlmFFRGuxzIGS3GaTKYzdzNYHHgO2HhgTfX20VjKL0p1Ldp8yEHHxyYYV8MnpmUHZW63WTUAN62d0ociwiF\/kkrvq\/3J9VEV+t+j0l4osN0kA0vU66Spezpzl\/0W+Hy9vn9QVUujhkFJKauCFIqDU4YQ==';

$signingSecret = 'ENV10VjX3YfNUIWZLeiRMDhXU_PucAVzmTYpC9UHbrr92M9JY_mfKRNmherFPyCwnZ4KKuUTscHzuDMh';

$computedSignature = hash_hmac('sha256', $request->getContent(), $signingSecret);

return hash_equals($signature, $computedSignature);

What you're doing basically is saying ok, if I take my webhook client secret value, and I use that to hash the content that has been sent over by PayPal, does the resulting value match the signature that PayPal sent as the paypal-transmission-sig header value.

Does that make more sense?

**Reply by Richard2012** [reply]
It makes perfect sencs, thank you for the reply in rocket time, will be back tomorrow.
Should the real values I have posted return true at verification time?

**Reply by PaulUniHomes** [reply]
Yes, but that depends and changes based on the content that comes over.

**Reply by Richard2012** [reply]
I have now removed the quotation marks from WEBHOOK_CLIENT_SECRET and it works. It was that stupid.
